text/0000-match-iter.md

```diff
@@ -0,0 +1,45 @@
+- Feature Name: match-iter
+- Start Date: 2017-07-13
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+Add the ability to continue a regular expression match throughout a subject string multiple times. This will allow developers to run through matching the same regex against the same target in a for loop, to collect results into an array, etc.
+
+# Motivation
+
+Not all regular expressions can know the full extent or the total number of groups they're planning on capturing. If you want to capture the same pattern multiple times within a single subject string, today developers have to write their own iterator or while loop. Languages like `Rust` have an iterator built into their regular expression functionality and we would like Pony to have parity in this area.
+
+# Detailed design
+
+The implementation of this would be an addition to the standard library. We want to be able to jump from a validated regular expresion (e.g. the result of an `apply` on a `Regex`) to an iterator. One way to to this would be to add a `values()` method to `Match` that returns `Iterator[Match]`. 
+
+By implementing this iterator in the standard library, developers will have access to this functionality in a standardized, reliable fashion. Here is an example of how this might be used where a single (first) `Match` is converted into an iterator via the idiomatic `values()` method:
+
+```pony
+let r = Regex("([+-]?\\s*\\d+[dD]\\d+|[+-]?\\s*\\d+)")
+for m in r("3d6+10+1d20").values() do
```

Comment:
Looking at this code snippet, I'd actually expect it to be doing something other than what you've described.

Since `r("3d6+10+1d20")` returns a single `Match` object, that means you're calling `Match.values()` here, which implies to me that you'll be iterating over each of the capture groups that are accessible via `Match.apply(index: USize)`. That is, I'd expect it to be an `Iterator[String]` where each `String` is a capture group, instead of an `Iterator[Match]` where each `Match` is a subsequent match in the original string, as you've laid out in the RFC.

What I think we probably want is something like this:

```pony
let regex = Regex("([+-]?\\s*\\d+[dD]\\d+|[+-]?\\s*\\d+)")
or m in regex.matches("3d6+10+1d20") do
  // ...
end
```